From: Branden Robinson <branden@debian.org>
Subject: add support for logfile rotation

- Implement new ReopenLogFile() and ReopenLogFileNotify() functions.
  ReopenLogFileNotify() is a signal handler for SIGUSR2.
- Move the definition of the WRITES() macro from error.c to dm_error.h
  so that dm.c can use it as well.
- Document xdm's signal handling in its manpage.

Forward-ported by Eugene Konev and Julien Cristau.

Index: xdm/dm.c
===================================================================
--- xdm.orig/dm.c
+++ xdm/dm.c
@@ -83,8 +83,10 @@
 extern FILE    *fdopen();
 #endif
 
-static SIGVAL	StopAll (int n), RescanNotify (int n);
+static SIGVAL	StopAll (int n), RescanNotify (int n),
+				ReopenLogFileNotify (int n);
 static void	RescanServers (void);
+static void ReopenLogFile (void);
 static void	RestartDisplay (struct display *d, int forceReserver);
 static void	ScanServers (void);
 static void	SetAccessFileTime (void);
@@ -93,6 +95,7 @@
 static void	TerminateProcess (pid_t pid, int signal);
 
 volatile int	Rescan;
+volatile int	Reopen;
 static long	ServersModTime, ConfigModTime, AccessFileModTime;
 
 int nofork_session = 0;
@@ -202,6 +205,7 @@
     AddOtherEntropy();
 #endif
     (void) Signal (SIGHUP, RescanNotify);
+	(void) Signal (SIGUSR2, ReopenLogFileNotify);
 #ifndef UNRELIABLE_SIGNALS
     (void) Signal (SIGCHLD, ChildNotify);
 #endif
@@ -212,6 +216,11 @@
 #endif
 	   AnyDisplaysLeft ())
     {
+	if (Reopen)
+	{
+	    ReopenLogFile ();
+		Reopen = 0;
+	}
 	if (Rescan)
 	{
 	    RescanServers ();
@@ -236,6 +245,7 @@
     int olderrno = errno;
 
     Debug ("Caught SIGHUP\n");
+	Reopen = 1;
     Rescan = 1;
 #ifdef SIGNALS_RESET_WHEN_CAUGHT
     (void) Signal (SIGHUP, RescanNotify);
@@ -243,6 +253,26 @@
     errno = olderrno;
 }
 
+/*
+ * Handle a SIGUSR2: set variable that will instruct the main loop to
+ * reopen the log file.
+ */
+static void
+ReopenLogFileNotify (int n)
+{
+#ifdef SIGNALS_RESET_WHEN_CAUGHT
+    int olderrno = errno;
+#endif
+
+    /* Debug() is not safe inside a signal handler. */
+    WRITES(STDERR_FILENO, "ReopenLogFileNotify handling SIGUSR2\n");
+    Reopen = 1;
+#ifdef SIGNALS_RESET_WHEN_CAUGHT
+    (void) Signal (SIGUSR2, ReopenLogFileNotify);
+    errno = olderrno;
+#endif
+}
+
 static void
 ScanServers (void)
 {
@@ -310,6 +340,14 @@
 }
 
 static void
+ReopenLogFile (void)
+{
+    Debug ("closing standard error file descriptor %d\n", STDERR_FILENO);
+    close (STDERR_FILENO);
+    InitErrorLog ();
+}
+
+static void
 SetConfigFileTime (void)
 {
     struct stat	statb;
Index: xdm/dm_error.h
===================================================================
--- xdm.orig/dm_error.h
+++ xdm/dm_error.h
@@ -44,6 +44,8 @@
 # define GCC_PRINTFLIKE(fmt,var) /*nothing*/
 #endif
 
+#define WRITES(fd, buf) write(fd, buf, strlen(buf))
+
 extern void Debug        (char * fmt, ...) GCC_PRINTFLIKE(1,2);
 extern void InitErrorLog (void);
 extern void LogError     (char * fmt, ...) GCC_PRINTFLIKE(1,2);
Index: xdm/error.c
===================================================================
--- xdm.orig/error.c
+++ xdm/error.c
@@ -49,8 +49,6 @@
 #include "dm.h"
 #include "dm_error.h"
 
-#define WRITES(fd, buf) write(fd, buf, strlen(buf))
-
 void
 LogInfo(char * fmt, ...)
 {
Index: xdm/xdm.man.cpp
===================================================================
--- xdm.orig/xdm.man.cpp
+++ xdm/xdm.man.cpp
@@ -1411,6 +1411,37 @@
 multiple window systems on the same hardware, you'll probably be more
 interested in
 .I xinit.
+.SH "ASYNCHRONOUS EVENTS"
+.B xdm
+uses
+.B SIGALRM
+and
+.B SIGUSR1
+for its own inter-process communication purposes, managing the relationship
+between the parent
+.B xdm
+process and its children.
+Sending these signals to any
+.B xdm
+process may result in unexpected behavior.
+.TP
+.B SIGHUP
+causes
+.B xdm
+to rescan its configuration files and reopen its log file.
+.TP
+.B SIGTERM
+causes
+.B xdm
+to terminate its children and shut down.
+.TP
+.B SIGUSR2
+causes
+.B xdm
+to reopen its log file.
+This is useful if log rotation is desired, but
+.B SIGHUP
+is too disruptive.
 .SH FILES
 .TP 20
 .I XDMDIR/xdm-config
